mxfusion/modules/gp_modules/dgp.py

```diff
@@ -0,0 +1,479 @@
+        kern_params = kern.fetch_parameters(variables)
+
+        # Compute covariance matrix from parts
+        cov = F.linalg.syrk(S_W) + make_diagonal(F, S_diag)
+
+        _, M, D = mu.shape
+        k_uu = kern.K(F, z, **kern_params) + F.eye(M, dtype=dtype) * jitter
+        L = F.linalg.potrf(k_uu)
+        Kff_diag = kern.Kdiag(F, x, **kern_params)
+
+        Ls = F.linalg.potrf(cov)
+        samples, Ls, z_s, Kff_diag = arrays_as_samples(F, [samples, Ls, z, Kff_diag])
+
+        k_xu = kern.K(F, samples, z_s, **kern_params)
+
+        # find q(f) from q(u)
+        mu_f, v_f = compute_marginal_f(F, L, k_xu, Kff_diag, mu, Ls)
+
+        # sample from q(f)
+        L_f = F.sqrt(v_f)
+        die = random_generator.sample_normal(shape=(n_samples, x.shape[1], D), dtype=dtype)
+
+        samples = F.broadcast_add(F.broadcast_mul(L_f, die), mu_f)
+
+    return samples, mu_f, v_f
+
+
+class DeepGPLogPdf(VariationalInference):
+    """
+    The inference algorithm for computing the variational lower bound of the stochastic variational Gaussian process
+    with Gaussian likelihood.
+    """
+
+    def __init__(self, model, posterior, observed, n_layers, jitter=1e-6, n_samples=10, dtype=None):
+        """
+        :param model: Model graph
+        :param posterior: Posterior graph
+        :param observed: List of observed variables
+        :param jitter: Jitter for numerical stability, defaults to 1e-6
+        """
+
+        super().__init__(model=model, posterior=posterior, observed=observed)
+
+        if dtype is None:
+            self.dtype = get_default_dtype()
+        else:
+            self.dtype = dtype
+
+        self.log_pdf_scaling = 1.
+        self.jitter = jitter
+        self.n_layers = n_layers
+        if self.n_layers <= 1:
+            raise ValueError('Number of layers must be greater than one')
+        self._rand_gen = MXNetRandomGenerator
+        self.n_samples = n_samples
+
+    def compute(self, F, variables):
+        """
+        Compute ELBO of model
+
+        :param F:
+        :param variables:
+        :return:
+        """
+        y = variables[self.model.Y]
+
+        # Compute model fit term
+        _, mu_f, v_f = dgp_prediction(F, variables, self.model, self.posterior, self.n_layers,
+                                      self._rand_gen, self.n_samples, self.jitter, self.dtype)
+        noise_var = variables[self.model.noise_var]
+
+        lik = gaussian_variational_expectation(F, y, noise_var, mu_f, v_f)
+
+        # Compute kl term
+        kl = self._compute_kl(F, variables)
+        return self.log_pdf_scaling * lik.mean(axis=0).sum() - kl
+
+    def _compute_kl(self, F, variables):
+        """
+        Compute sum of KL divergences for each layer of DGP
+        """
+        kl = 0
+        for layer in range(self.n_layers):
+            # Get layer specific variables
+            z = variables[getattr(self.model, 'inducing_inputs_' + str(layer))]
```

Comment:
Hey I know I suggested doing this "getattr" logic,  but I think a better way would just be to implement a simple getter on the FactorGraph. I've made an issue for it #167. If you feel like doing this in this CR that'd be great.